[tp] enable rowwise embedding sharding in RowwiseParallel

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118330
* __->__ #118242
* #118080
* #118079

As titled, this PR enables the rowwise embedding sharding in the
RowwiseParallel style, and add tests to ensure it's working as expected

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[tp] enable rowwise embedding sharding in RowwiseParallel

As titled, this PR enables the rowwise embedding sharding in the
RowwiseParallel style, and add tests to ensure it's working as expected

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -179,10 +179,6 @@ def __init__(
         super().__init__()
         self.input_layouts = (input_layouts or Shard(-1), )
```
Now we have two use cases nn.Linear and nn.Embedding, with two different `desired_input_layouts`. In the long term does it still make sense to keep default `input_layouts` to be `Shard(-1)`? This seems favoring nn.Linear use cases. In the short term, we need to keep it for backward compatibility.

wanchaol replied:
That's a good question. Yeah for BC reasons I think we probably can't change the default input layouts for rowwise parallel, and I think we should indeed favoring nn.Linear instead of nn.Embedding as Embedding only happen in the first transformer layer most of time.

I think currently user have to pass `input_layouts=Replicate()` for models like Transoformer, one thing we can do is that maybe we should add a warning if we found that we are running row sharded embedding and warn user if the input_layouts is not replicate?

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -195,32 +191,51 @@ def _prepare_input_fn(input_layouts, desired_input_layouts, inputs, device_mesh)
             module,
             device_mesh,
-            self._partition_fn,
+            partition_fn,
```
The issue is mainly distribute_module needs to know desired_input_layouts when calling `_apply`, so we have to do this in the beginning of _apply, I'll refactor Colwise to make them aligned

wanchaol committed:
Update on "[tp] enable rowwise embedding sharding in RowwiseParallel"


As titled, this PR enables the rowwise embedding sharding in the
RowwiseParallel style, and add tests to ensure it's working as expected

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -195,32 +193,51 @@ def _prepare_input_fn(input_layouts, desired_input_layouts, inputs, device_mesh)
+        if isinstance(module, nn.Linear):
+            partition_fn = self._partition_linear_fn
+            # rowwise linear runtime sharding requires input tensor shard on last dim
+            self.desired_input_layouts: Tuple[Placement, ...] = (Shard(-1), )
+        elif isinstance(module, nn.Embedding):
+            partition_fn = self._partition_embedding_fn
+            # rowwise embedding runtime sharding requires input tensor replicated
+            self.desired_input_layouts = (Replicate(), )
```
up to you whether or not we want L229 and L233 more symmetric

wanchaol replied:
If you are referring to type annotation then mypy would complain if I declare the type twice.

tianyu-l reviewed (approved):
ready to merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
cc @joshlk as we discussed the rowwise embedding sharding a while ago, now it's added!

wanchaol commented:
@DanilBaibak Given that the reverts was due to the ghexport failures in the internal, not related to any internal tests or OSS tests, I think it's a dev infra issue rather than the PR issue, I'll try to reland them directly

wanchaol commented:
@pytorchbot merge
